sktime/forecasting/model_selection/tests/test_tune.py

```diff
@@ -348,3 +352,57 @@ def test_gscv_backends(backend_set):
         backend_params=backend_params,
     )
     gscv.fit(y, X)
+
+
+TEST_PARAMS_DICT = PIPE_GRID
+
+TEST_PARAMS_LIST = [
+    {
+        "window_length": [1, 2, 3],
+        "strategy": ["last", "mean"],
+        "transformer__degree": [1, 2, 3],
+        "forecaster__strategy": ["last", "mean", "seasonal_last"],
+    },
+    {
+        "window_length": [4, 5, 6],
+        "forecaster__strategy": ["last", "mean"],
+    },
+]
+
+
+@pytest.mark.parametrize("return_n_best_forecasters", [-1, 0, 3])
+@pytest.mark.parametrize(
+    "Forecaster, kwargs",
+    [
+        (ForecastingGridSearchCV, {"param_grid": TEST_PARAMS_DICT}),
+        (ForecastingGridSearchCV, {"param_grid": TEST_PARAMS_LIST}),
+        (ForecastingRandomizedSearchCV, {"param_distributions": TEST_PARAMS_LIST}),
+    ],
+)
+def test_return_n_best_forecasters(Forecaster, return_n_best_forecasters, kwargs):
+    y, X = load_longley()
+    searchCV = Forecaster(
+        forecaster=PIPE,
+        cv=CVs[0],
+        **kwargs,
+        return_n_best_forecasters=return_n_best_forecasters,
+    )
+    searchCV.fit(y, X)
+    if return_n_best_forecasters == -1:
+
+        def calculate_total_combinations(param_grid: Union[List[Dict], Dict]):
+            if isinstance(param_grid, dict):
+                return reduce(lambda x, y: x * y, [len(x) for x in param_grid.values()])
+            elif isinstance(param_grid, list):
+                return sum(calculate_total_combinations(i) for i in param_grid)
+            else:
+                error_message = "`param_grid` must be a dict or a list[dict]"
+                raise ValueError(error_message)
+
+        if "param_grid" in kwargs:
+            total_combinations = calculate_total_combinations(kwargs["param_grid"])
+            assert len(searchCV.n_best_forecasters_) == total_combinations
+        else:
+            assert len(searchCV.n_best_forecasters_) == searchCV.n_iter
```

Comment:
Just realised, this may not be true in general. We instead should check against minimum of n_iter and cardinality of the parameter space for the case where there are not many points in the parameter grid as the specified number of iterations to sample from it.

Reply:
I was just looking into this, and there is another problem. If `n_iter` is larger than the total search space, it gives a **warning**. But when `return_n_best_forecasters` is larger than the search space, it gives an **error**.

Is this behavior intended because it looks inconsistent.

Reply:
This is the error `IndexError: single positional indexer is out-of-bounds`

This is an example code almost the same as the docstring example [here](https://github.com/sktime/sktime/blob/2531edb6da3e6a6ef71bc27664ea6a75f0dadf4d/sktime/forecasting/model_selection/_tune.py#L536-L549)

```py
from sktime.datasets import load_shampoo_sales
from sktime.forecasting.model_selection import ForecastingRandomizedSearchCV
from sktime.split import ExpandingWindowSplitter
from sktime.forecasting.naive import NaiveForecaster
y = load_shampoo_sales()
fh = [1, 2, 3]
cv = ExpandingWindowSplitter(fh=fh)
forecaster = NaiveForecaster()
param_grid = {"strategy": ["last", "mean", "drift"]}
gscv = ForecastingRandomizedSearchCV(
     forecaster=forecaster,
     param_distributions=param_grid,
     return_n_best_forecasters=4,
     cv=cv
     )
gscv.fit(y)
```

I think [this](https://github.com/sktime/sktime/blob/2531edb6da3e6a6ef71bc27664ea6a75f0dadf4d/sktime/forecasting/model_selection/_tune.py#L271) is the exact line causing the problem

Reply:
Please check https://github.com/sktime/sktime/pull/6031#discussion_r1521836163.